path/src/util/compact_bytes.rs

```diff
@@ -0,0 +1,164 @@
+// MIT LICENSE
+//
+// Copyright (c) 2023 Dash Core Group
+//
+// Permission is hereby granted, free of charge, to any
+// person obtaining a copy of this software and associated
+// documentation files (the "Software"), to deal in the
+// Software without restriction, including without
+// limitation the rights to use, copy, modify, merge,
+// publish, distribute, sublicense, and/or sell copies of
+// the Software, and to permit persons to whom the Software
+// is furnished to do so, subject to the following
+// conditions:
+//
+// The above copyright notice and this permission notice
+// shall be included in all copies or substantial portions
+// of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+// ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT
+// SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
+// IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! Compact two dimenstional bytes array structure.
+
+use std::mem;
+
+/// Bytes vector wrapper to have multiple byte arrays allocated continuosuly.
+#[derive(Debug, Default)]
+pub(crate) struct CompactBytes {
+    n_segments: usize,
+    data: Vec<u8>,
+}
+
+impl CompactBytes {
+    /// Create empty structure.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Append bytes segment.
+    pub fn add_segment(&mut self, segment: &[u8]) {
+        // Because the iteration will go backwards, a segment is inserted an odd way:
+        // First it goes the segment itself
+        self.data.extend_from_slice(segment);
+        // Next it will be it's length
+        self.data.extend_from_slice(&segment.len().to_ne_bytes());
```

Comment:
shouldn't the len be first?

Also I'm not a fan of ne bytes because things can be different depending on the machine, be_bytes would be better here I believe.

Reply:
It's because we'll iterate from the end, we know the end (vec's length) and just pick last sizeof(usize) bytes to get length of a segment to go back;

If we decided to have segment length first, we had to store "last segment length" position, then read it and bytes after, but how we get length of the previous segment?

This data won't leave process' memory anyway